python/singa/sonnx.py

```diff
@@ -0,0 +1,270 @@
+            k = (int(i.attribute[0].ints[0]),int(i.attribute[0].ints[0]))
+            layer[str(i.output[0])] = autograd.AvgPool2d(k, int(i.attribute[2].ints[0]),padding=int(i.attribute[1].ints[0]))
+    for i in model.graph.node:
+        if (i.op_type == 'BatchNormalization'):
+            shape = find_shape(i.input[1], model)
+            layer[str(i.output[0])] = autograd.BatchNorm2d(shape[0])
+            layer[str(i.output[0])].set_params(scale=modeldic[str(i.input[1])])
+            layer[str(i.output[0])].set_params(bias=modeldic[str(i.input[2])])
+
+    return layer, model
+
+class ONNXm(Layer):
+    def __init__(self,path):
+        super(ONNXm, self).__init__()
+        self.modeldic, self.model = onnx_model_init(path)
+        self.layer, self.model = combine_node(self.modeldic, self.model)
+
+    def __call__(self,inputs):
+        '''
+            input:
+            a graph node dictionary
+            model: graph model
+            target: label
+
+            load other nodes of onnx
+            '''
+        layer, model,oper = self.layer, self.model,self.modeldic
+        self.modeldic['X'] = inputs
+        for i in model.graph.node:
+            if (i.op_type == 'Constant'):
+                pass
+                # do nothing
+            if (i.op_type == 'Relu'):
+                oper[str(i.output[0])] = autograd.relu(oper[str(i.input[0])])
+            elif (i.op_type == 'Softmax'):
+                oper[str(i.output[0])] = autograd.softmax(oper[str(i.input[0])])
+            elif (i.op_type == 'Add'):
+                oper[str(i.output[0])] = autograd.add(oper[str(i.input[0])], oper[str(i.input[1])])
+            elif (i.op_type == 'MatMul'):
+                oper[str(i.output[0])] = autograd.matmul(oper[str(i.input[0])], oper[str(i.input[1])])
+            elif (i.op_type == 'Linear'):
+                oper[str(i.output[0])] = layer[str(i.output[0])](oper[str(i.input[0])])
+            elif (i.op_type == 'Conv'):
+                #print(tensor.to_numpy(oper[str(i.input[0])]).shape)
+                #print(i.output[0])
+                oper[str(i.output[0])] = layer[str(i.output[0])](oper[str(i.input[0])])
+            elif (i.op_type == 'Flatten'):
+                oper[str(i.output[0])] = autograd.flatten(oper[str(i.input[0])])
+            elif(i.op_type == 'Concat'):
+                oper[str(i.output[0])] = autograd.cat((oper[str(i.input[0])], oper[str(i.input[1])]),int(i.attribute[0].i))
+            elif(i.op_type == 'Tanh'):
+                oper[str(i.output[0])] = autograd.tanh(oper[str(i.input[0])])
+            elif (i.op_type == 'Sigmoid'):
+                oper[str(i.output[0])] = autograd.sigmoid(oper[str(i.input[0])])
+            elif (i.op_type == 'Mul'):
+                oper[str(i.output[0])] = autograd.mul(oper[str(i.input[0])],oper[str(i.input[1])])
+            elif (i.op_type == 'MaxPool'):
+                oper[str(i.output[0])] = layer[str(i.output[0])](oper[str(i.input[0])])
```

Comment:
since the MaxPool, Conv, BatchNorm have already been combined and converted into layer, can we replace the conditional check, i.e., `elif` by using the layer directly from `layer` dictionary.

Reply:
yes